### PR TITLE
Add Kafka headers as column

### PR DESCRIPTION
**Pull request by 0xE282B0**
This PR adds the internal _headers field to the kafka connector.
As type a map with key String and value array of byte[] is used.

For better review the changes are splited into two commits:
1.  Extract the type definitions from the enum that describes the internal fields to a map that is initialized in the constructor where the TypeManager can be used.
1. Add the Header column and adapt the KafkaRecordCursor

I would appreciate if someone could review the code.

### CODE REVIEW

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java [hunk 1]
```diff
@@ -134,6 +139,22 @@ private void insertData(byte[] data)
         }
     }
 
+    private void createMessagesWithHeader(String topicName)
+    {
+        try (KafkaProducer<byte[], byte[]> producer = createProducer()) {
+            ProducerRecord<byte[], byte[]> record = new ProducerRecord<byte[], byte[]>(topicName, null, "{}".getBytes(Charset.forName("UTF-8")));
+            record.headers()
+                    .add("notfoo", null);
```
Comment (aalbu) [inline presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:147]: Can you change the value to something other than `null`?  As the test is written, it's not clear whether the `null` in the result comes from this header or the one a few lines below (`.add("foo", null)`).

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java [hunk 2]
```diff
@@ -269,4 +279,58 @@ public void close()
...
+            @Override
+            public boolean isNull()
+            {
+                return !headers.iterator().hasNext();
+            }
+
+            @Override
+            public Block getBlock()
+            {
+                return (Block) varcharMapType.getObject(mapBlockBuilder, 0);
```
Comment (aalbu) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:332]: You don't need the cast.

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java [hunk 3]
```diff
@@ -13,126 +13,191 @@
...
+        {
+            return columnName;
+        }
     }
 
-    public String getColumnName()
+    /**
+     * Inner class that wraps the {@link InternalFields} enum and add {@link Type}.
+     * With this bean a non static data structure with types from the {@link TypeManager} can be created.
+     */
```
Comment (aalbu) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java:126]: I don't know if this is fit for a Javadoc comment.

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java [hunk 4]
```diff
@@ -134,6 +139,22 @@ private void insertData(byte[] data)
...
+            producer.send(record);
+            record = new ProducerRecord<byte[], byte[]>(topicName, null, "{}".getBytes(Charset.forName("UTF-8")));
+            record.headers()
+                    .add("foo", "bar".getBytes(Charset.forName("UTF-8")))
+                    .add("foo", null)
+                    .add("foo", "baz".getBytes(Charset.forName("UTF-8")));
+            producer.send(record);
```
Comment (aalbu) [inline presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:154]: Can you also publish a message without headers?

File: presto-docs/src/main/sphinx/connector/kafka.rst [hunk 5]
```diff
@@ -146,21 +146,22 @@ Internal Columns
...
-``_message_corrupt``    BOOLEAN   True if the decoder could not decode the message for this row. When true, data columns mapped from the message should be treated as invalid.
-``_message``            VARCHAR   Message bytes as an UTF-8 encoded string. This is only useful for a text topic.
-``_message_length``     BIGINT    Number of bytes in the message.
-``_key_corrupt``        BOOLEAN   True if the key decoder could not decode the key for this row. When true, data columns mapped from the key should be treated as invalid.
-``_key``                VARCHAR   Key bytes as an UTF-8 encoded string. This is only useful for textual keys.
-``_key_length``         BIGINT    Number of bytes in the key.
-======================= ========= =============================
+======================= =============================== =============================
+Column name             Type                            Description
+======================= =============================== =============================
```
Comment (findepi) [inline presto-docs/src/main/sphinx/connector/kafka.rst:151]: Please make the table wider in a separate commit, as in https://github.com/prestosql/presto/pull/3771/commits

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java [hunk 6]
```diff
@@ -269,4 +279,58 @@ public void close()
...
     }
+
+    public static FieldValueProvider headerMapValueProvider(MapType varcharMapType, Headers headers)
+    {
+        final Type keyType = varcharMapType.getTypeParameters().get(0);
+        final Type valueArrayType = varcharMapType.getTypeParameters().get(1);
+        final Type valueType = valueArrayType.getTypeParameters().get(0);
+
+        final BlockBuilder mapBlockBuilder = varcharMapType.createBlockBuilder(null, 1);
+        final BlockBuilder builder = mapBlockBuilder.beginBlockEntry();
```
Comment (findepi) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:290]: Per code style, we do not use `final` for local variables.
For short methods, it doesn't add to readability, and we strive to avoid long methods.

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java [hunk 7]
```diff
@@ -269,4 +279,58 @@ public void close()
...
+            builder.closeEntry();
+        }
+
+        mapBlockBuilder.closeEntry();
+
+        return new FieldValueProvider() {
+            @Override
+            public boolean isNull()
+            {
+                return !headers.iterator().hasNext();
```
Comment (findepi) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:326]: Is there a value in returning NULL when no headers?
Why not return empty map?

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java [hunk 8]
```diff
@@ -134,6 +139,22 @@ private void insertData(byte[] data)
         }
     }
 
+    private void createMessagesWithHeader(String topicName)
+    {
+        try (KafkaProducer<byte[], byte[]> producer = createProducer()) {
+            ProducerRecord<byte[], byte[]> record = new ProducerRecord<byte[], byte[]>(topicName, null, "{}".getBytes(Charset.forName("UTF-8")));
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:145]: this, + static import `UTF_8`

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java [hunk 9]
```diff
@@ -171,6 +192,17 @@ public void testRoundTripAllFormats(RoundTripTestCase testCase)
...
     }
 
+    @Test
+    public void testKafkaHeaders()
+    {
+        createMessagesWithHeader(headersTopic);
+
+        assertQuery("SELECT from_utf8(value) FROM default." + headersTopic +
+                        " CROSS JOIN UNNEST(_headers['foo']) as arr (value)" +
+                        " WHERE element_at(_headers,'foo') is not NULL",
```
Comment (findepi) [inline presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:202]: nit

```suggestion
                        " WHERE element_at(_headers, 'foo') IS NOT NULL",
```

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java (hunk 6 above)
Comment (0xE282B0) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java]: Thats fine.

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java [hunk 10]
```diff
@@ -269,4 +279,58 @@ public void close()
...
+            }
+            else {
+                final List<byte[]> list = new ArrayList<>();
+                list.add(header.value());
+                headerMap.put(key, list);
+            }
+        }
+
+        for (Map.Entry<String, List<byte[]>> header : headerMap.entrySet()) {
+            keyType.writeSlice(builder, utf8Slice(header.getKey()));
```
[the remark this answers is not in the document]
Comment (0xE282B0) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java]: Thanks, didn't know that.

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java (hunk 7 above)
Comment (0xE282B0) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java]: I just liked `SELECT _headers FROM topic1 WHERE _headers IS NOT NULL;` better than `SELECT _headers FROM topic1 WHERE cardinality(_headers) > 0;` but thats not a strong opinion. In fact you are right that the headers are empty and not absent.

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java [hunk 11]
```diff
@@ -54,13 +55,16 @@
...
     private String rawFormatTopic;
+    private String headersTopic;
 
     @Override
     protected QueryRunner createQueryRunner()
             throws Exception
     {
         testingKafka = new TestingKafka();
         rawFormatTopic = "test_raw_" + UUID.randomUUID().toString().replaceAll("-", "_");
+        headersTopic = "header_test" + UUID.randomUUID().toString().replaceAll("-", "_");
```
[the remark this answers is not in the document]
Comment (0xE282B0) [reply presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java]: sure, missed to copy.

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java [hunk 12]
```diff
@@ -76,6 +80,7 @@ protected QueryRunner createQueryRunner()
                                         createOneFieldDescription("boolean_int", BooleanType.BOOLEAN, "41", "INT"),
                                         createOneFieldDescription("boolean_short", BooleanType.BOOLEAN, "45", "SHORT"),
                                         createOneFieldDescription("boolean_byte", BooleanType.BOOLEAN, "47", "BYTE")))))
+                .put(new SchemaTableName("default", headersTopic), new KafkaTopicDescription(headersTopic, Optional.empty(), headersTopic, Optional.empty(), Optional.empty()))
```
[the remark this answers is not in the document]
Comment (0xE282B0) [reply presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java]: sure.

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (0xE282B0) [reply presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java]: better, thanks.

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java (hunk 1 above)
Comment (0xE282B0) [reply presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java]: good point.

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java (hunk 4 above)
Comment (0xE282B0) [reply presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:159]: good point.

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java (hunk 9 above)
Comment (0xE282B0) [reply presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java]: right.

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java (hunk 3 above)
Comment (0xE282B0) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java:126]: Should I write a non-doc comment instead or just remove it?

File: presto-docs/src/main/sphinx/connector/kafka.rst (hunk 5 above)
Comment (0xE282B0) [reply presto-docs/src/main/sphinx/connector/kafka.rst:151]: 👍

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java (hunk 3 above)
Comment (aalbu) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java:126]: I would opt for removal.

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java [hunk 13]
```diff
@@ -13,126 +13,187 @@
...
+        {
+            InternalFields description = BY_COLUMN_NAME.get(columnName);
+            checkArgument(description != null, "Unknown internal column name %s", columnName);
+            return description;
+        }
+
+        private final String columnName;
+        private final String comment;
+
+        InternalFields(
```
Comment (charlesjmorgan) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java:108]: Can just be one line

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java [hunk 14]
```diff
@@ -13,126 +13,187 @@
...
     {
-        checkArgument(!isNullOrEmpty(columnName), "name is null or is empty");
-        this.columnName = columnName;
-        this.type = requireNonNull(type, "type is null");
-        this.comment = requireNonNull(comment, "comment is null");
+        private final String columnName;
+        private final Type type;
+        private final String comment;
+
+        InternalFieldDescription(
```
Comment (charlesjmorgan) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java:129]: Can be one line

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java [hunk 15]
```diff
@@ -13,126 +13,187 @@
...
-                getType(),
-                null,
-                null,
-                null,
-                false,
-                hidden,
-                true);
+        Type varcharMapType = typeManager.getType(mapType(VARCHAR.getTypeSignature(), arrayType(VARBINARY.getTypeSignature())));
+
+        internalFields = new ImmutableMap.Builder<InternalFields, InternalFieldDescription>()
```
Comment (charlesjmorgan) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java:179]:
```suggestion
        internalFields = ImmutableMap.<InternalFields, InternalFieldDescription>builder()
```

Comment (charlesjmorgan) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java:179]:
```suggestion
        internalFields = ImmutableMap.<InternalFields, InternalFieldDescription>builder()
```

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java [hunk 16]
```diff
@@ -171,6 +197,17 @@ public void testRoundTripAllFormats(RoundTripTestCase testCase)
                 "VALUES (" + testCase.getFieldValues() + "), (" + testCase.getFieldValues() + ")");
     }
 
+    @Test
+    public void testKafkaHeaders()
```
Comment (charlesjmorgan) [inline presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:201]: Move this before `testRoundTripAllFormats` so that it doesn't split that test from its data provider

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java (hunk 15 above)
Comment (0xE282B0) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java]: thanks.

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java [hunk 17]
```diff
@@ -111,19 +118,13 @@ public String getColumnName()
         }
     }
 
-    /**
-     * Inner class that wraps the {@link InternalFields} enum and add {@link Type}.
-     * With this bean a non static data structure with types from the {@link TypeManager} can be created.
```
Comment (findepi) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java]: Remove the comment in the previous commit, which was adding it

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java [hunk 18]
```diff
@@ -269,4 +278,53 @@ public void close()
...
+            String key = header.key();
+            if (headerMap.containsKey(key)) {
+                headerMap.get(key).add(header.value());
+            }
+            else {
+                List<byte[]> list = new ArrayList<>();
+                list.add(header.value());
+                headerMap.put(key, list);
+            }
+        }
```
Comment (findepi) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:303]: Use Multimap. 

```suggestion
        Multimap<String, byte[]> headerMap = ArrayListMultimap.create();
        for (Header header : headers) {
            headerMap.put(header.key(), header.value());
        }
```

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java [hunk 19]
```diff
@@ -269,4 +278,53 @@ public void close()
             kafkaConsumer.close();
         }
     }
+
+    public static FieldValueProvider headerMapValueProvider(MapType varcharMapType, Headers headers)
```
Comment (findepi) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:282]: I'd assume empty headers case is not a rare case, so we could want to cache a result for this (empty block).

@dain wdyt?

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java [hunk 20]
```diff
@@ -134,6 +139,27 @@ private void insertData(byte[] data)
...
+            // Message with simple header
+            record = new ProducerRecord<>(topicName, null, "{}".getBytes(UTF_8));
+            record.headers()
+                    .add("notfoo", "some value".getBytes(UTF_8));
+            producer.send(record);
+            // Message with multiple same key headers
+            record = new ProducerRecord<>(topicName, null, "{}".getBytes(UTF_8));
+            record.headers()
+                    .add("foo", "bar".getBytes(UTF_8))
+                    .add("foo", null)
```
Comment (findepi) [inline presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:157]: thanks for taking care of null values
can header key be null as well?

Comment (0xE282B0) [reply presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:157]: No, at least the Kafka Java client throws an exception if you try to create a Header with null as key.
See: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/header/internals/RecordHeader.java#L32

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java [hunk 21]
```diff
@@ -29,110 +32,160 @@
...
+                .put(InternalFields.PARTITION_ID_FIELD, new InternalFieldDescription(InternalFields.PARTITION_ID_FIELD, BigintType.BIGINT))
+                .put(InternalFields.PARTITION_OFFSET_FIELD, new InternalFieldDescription(InternalFields.PARTITION_OFFSET_FIELD, BigintType.BIGINT))
+                .put(InternalFields.MESSAGE_CORRUPT_FIELD, new InternalFieldDescription(InternalFields.MESSAGE_CORRUPT_FIELD, BooleanType.BOOLEAN))
+                .put(InternalFields.MESSAGE_FIELD, new InternalFieldDescription(InternalFields.MESSAGE_FIELD, createUnboundedVarcharType()))
+                .put(InternalFields.MESSAGE_LENGTH_FIELD, new InternalFieldDescription(InternalFields.MESSAGE_LENGTH_FIELD, BigintType.BIGINT))
+                .put(InternalFields.KEY_CORRUPT_FIELD, new InternalFieldDescription(InternalFields.KEY_CORRUPT_FIELD, BooleanType.BOOLEAN))
+                .put(InternalFields.KEY_FIELD, new InternalFieldDescription(InternalFields.KEY_FIELD, createUnboundedVarcharType()))
+                .put(InternalFields.KEY_LENGTH_FIELD, new InternalFieldDescription(InternalFields.KEY_LENGTH_FIELD, BigintType.BIGINT))
```
[the remark this answers is not in the document]
Comment (0xE282B0) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldDescription.java]: https://github.com/prestosql/presto/pull/4462/files#diff-a4d914e9d2a3a44c167509b820255365R108
@findepi, did I get you right?

**Comment by 0xE282B0** [comment]
Thanks for the review @losipiuk,
I agree that it feels odd to have the static enum and a representation as class with type at runtime. Therefore I'd like to get rid of the enum. 
@findepi WDYT?

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldManager.java [hunk 22]
```diff
@@ -138,6 +149,10 @@ public KafkaInternalFieldManager(TypeManager typeManager)
                         MESSAGE_FIELD,
                         "Message text",
                         createUnboundedVarcharType()))
+                .put(HEADERS_FIELD, new InternalField(
+                        HEADERS_FIELD,
+                        "Headers of the message as map.",
```
Comment (losipiuk) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInternalFieldManager.java:154]: drop dot at the end :)

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java (hunk 19 above)
Comment (losipiuk) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:291]: @0xE282B0 can you address this one unless the assumption about the fact, that empty headers is a common thing, is false?

**Review by losipiuk — COMMENTED** [review]
Thanks looks good. Minor comments.

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java (hunk 19 above)
Comment (0xE282B0) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:325]: sure.

Comment (0xE282B0) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:325]: Is there an easier way to get an empty MapBlock than creating a MapType and using a MapBlockBuilder?

```
MapType mapType = new MapType(VarcharType.VARCHAR, new ArrayType(VarbinaryType.VARBINARY),
        MethodHandles.empty(methodType(Boolean.class, Block.class, int.class, long.class)),
        MethodHandles.empty(methodType(Boolean.class, Block.class, int.class, Block.class, int.class)),
        MethodHandles.empty(methodType(long.class, Object.class)),
        MethodHandles.empty(methodType(long.class, Object.class)));
BlockBuilder mapBlockBuilder = new MapBlockBuilder(mapType, null, 0);
mapBlockBuilder.beginBlockEntry();
mapBlockBuilder.closeEntry();
Block emptyMapBlock = mapType.getObject(mapBlockBuilder, 0);
```

https://github.com/prestosql/presto/pull/4462/commits/414ddcb4a71e0c6d42a9a7ea62b105bcaec0e8fe#diff-1938e94844afe6ab2ee40936951490d9R300-R308

Comment (losipiuk) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:324]: I am not aware. Looks good to me. 

Nit: can you rename `createEmptyMapBlockProvider` to `createEmptyHeadersFieldProvider` 
and `EMPTY_MAP_BLOCK_PROVIDER` to `EMPTY_HEADERS_FIELD_PROVIDER`

Comment (losipiuk) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:324]: Also please add a test when we are reading `_headers` column but headers map is empty.

**Review by losipiuk — COMMENTED** [review]
Thanks. 
One more request for a test and we are good to go :)

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java (hunk 19 above)
Comment (0xE282B0) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:324]: Naming sounds better - more specific.
The empty header test doesn't look as straightforward as I wanted it to be, but it works. Since the `assertQuery` does not work on maps and VARBINARY does not work with JSON, I converted the map type and then cast it to JSON to compare it with empty objects. 
```
SELECT cast(transform_values(_headers,(k,v)->transform(v,x->from_utf8(x))) AS JSON)
```
Recomendations welcome 😉

Comment (losipiuk) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:324]: I think the test could be a bit more readable if headers topic also defined named `id` field. It would make assertions simpler.
Then if messages with empty headers has `id` equal to `1` and `2`.
The assertion would look like:
```
 assertQuery("SELECT id FROM default." + headersTopic + " WHERE cardinality(_headers) = 0",
                "VALUES (1), (2)");
```

The other test query could be more readable too if you used `id` in `WHERE` clause.

WDYT?

Comment (0xE282B0) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:324]: Unfortunately we don't have an ID, but we can enumerate the messages based on the value of the message.
```
assertQuery("SELECT _message FROM default." + headersTopic + " WHERE cardinality(_headers) = 0",
               "VALUES ('1'),('2')");
```

Comment (losipiuk) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java:324]: I was thinking of extending the schema with `id` but using `_message` is totally fine too.

**Review by losipiuk — APPROVED** [review]

**Comment by losipiuk** [comment]
Thanks! I did not notice previously that you already made a change to the test :)
